ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+    final byte[] key = serializeKey(row.getColumnValue(SchemaUtil.ROWKEY_NAME_INDEX), dataSource);
+    final byte[] value = serializeRow(row, dataSource, config, serviceContext);
+
+    final String topicName = dataSource.getKafkaTopicName();
+
+    // for now, just create a new producer each time
+    final Producer<byte[], byte[]> producer = serviceContext
+        .getKafkaClientSupplier()
+        .getProducer(config.getProducerClientConfigProps());
+
+    producer.send(
+        new ProducerRecord<>(
+            topicName,
+            null,
+            row.getColumnValue(SchemaUtil.ROWTIME_INDEX),
+            key,
+            value
+        )
+    );
+
+    producer.close(Duration.ofSeconds(MAX_SEND_TIMEOUT_SECONDS));
+
+    return Optional.empty();
+  }
+
+  private GenericRow extractRow(
+      final InsertValues insertValues,
+      final StructuredDataSource<?> dataSource
+  ) {
+    final Optional<String> keyField = dataSource.getKeyField().name();
+
+    final List<String> columns = insertValues.getColumns().isEmpty()
+        ? dataSource.getSchema()
+          .fields()
+          .stream()
+          .map(Field::name)
+          .filter(name -> !SchemaUtil.ROWTIME_NAME.equals(name))
+          .collect(Collectors.toList())
+        : insertValues.getColumns();
+
+    if (columns.size() != insertValues.getValues().size()) {
+      // this will only happen if insertValues.getColumns() is empty, otherwise
+      // the columns/values are verified by the InsertValues constructor
+      throw new KsqlException(
+          "Expected a value for each column. Expected Columns: " + columns
+              + ". Got " + insertValues.getValues());
+    }
+
+    final Map<String, Object> values = new HashMap<>();
+    for (int i = 0; i < columns.size(); i++) {
+      final String column = columns.get(i);
+      final Schema columnSchema = dataSource.getSchema().field(column).schema();
+      final Expression valueExp = insertValues.getValues().get(i);
+
+      values.put(column, new ExpressionResolver(columnSchema, column).process(valueExp, null));
+    }
+
+    if (keyField.isPresent()) {
+      final String key = keyField.get();
+      final Object keyValue = values.get(key);
+      final Object rowKeyValue = values.get(SchemaUtil.ROWKEY_NAME);
+
+      if ((keyValue != null) ^ (rowKeyValue != null)) {
+        values.putIfAbsent(key, rowKeyValue);
+        values.putIfAbsent(SchemaUtil.ROWKEY_NAME, keyValue);
+      } else if (!Objects.equals(keyValue, rowKeyValue)) {
+        throw new KsqlException(
+            String.format(
+                "Expected ROWKEY and %s to match but got %s and %s respectively.",
+                key, rowKeyValue, keyValue));
+      }
+    }
+
+    values.putIfAbsent(SchemaUtil.ROWTIME_NAME, clock.getAsLong());
+
+    for (Field field : dataSource.getSchema().fields()) {
+      if (!field.schema().isOptional() && values.getOrDefault(field.name(), null) == null) {
```

Comment:
All fields are currently optional, right?  Is this just future proofing?

Reply:
yes